### ISSUE TITLE
Handful of `Dictionary`s somewhere to centralize "entity group" storage (one for each `Entity` type)

### ISSUE THREAD

**Issue by hamstar0** [tModLoader/tModLoader#1004]
### Description
See title. That would literally be as simple as something like:
```cs
public class ModContent {
	private static IDictionary<string, IReadOnlyList<int>> _ItemGroups;
	public static IReadOnlyDictionary<string, IReadOnlyList<int>> ItemGroups { get; private set; }

	static ModContent() {
		_ItemGroups = new ConcurrentDictionary<string, IReadOnlyList<int>>();
		ItemGroups = new ReadOnlyDictionary<string, IReadOnlyList<int>>( _ItemGroups );
	}

	public static void AddToItemGroup( string groupName, IEnumerable<int> group ) {
		if( !_ItemGroups.ContainsKey( groupName ) ) {
			_ItemGroups[groupName] = group.ToList().AsReadOnly();
		} else {
			_ItemGroups[groupName] = _ItemGroups[groupName].Union( group ).ToList().AsReadOnly();
		}
	}

	...
}
```

### What does this proposal attempt to solve or improve?
A place for mods to look for, and contribute to, listings of entities (by their `type` field's value) by a given name (e.g. any 'spooky' NPC, any explosive item, etc.).

**Comment by Chicken-Bones** [comment]
Basically ore dictionary from minecraft forge. Yes

**Comment by Mirsario** [comment]
From my experience with overhaul, constant string dictionary access on entities will slow things down incredibly. I would prefer just making a proper tag system (that I could use in overhaul) with OOP tag definitions that will actually just be shortcuts to generic static data, so that strings don't need to be used in contexts where performance matters. (Your usage would still be possible.)

**Comment by hamstar0** [comment]
> From my experience with overhaul, constant string dictionary access on entities will slow things down incredibly. I would prefer just making a proper tag system (that I could use in overhaul) with OOP tag definitions that will actually just be shortcuts to generic static data, so that strings don't need to be used in contexts where performance matters. (Your usage would still be possible.)

Could you give a code example? I'm unclear on the meaning of tags in this context.

Keep in mind part of this idea is that modders won't always know to assign their items, NPCs, etc. (entities) to groups in their mods. This method lets anyone assign any given entity (by its `type`) to any given group, and thus crowd sourced group definitions (I'm going to assume there won't be issues with this concept, since mods are already practically naked to one another to begin with).

**Comment by Mirsario** [comment]
@hamstar0 here's one I gave to CB: https://paste.mod.gg/jituketadi.cpp
Tags are the same thing you meant by groups. They're descriptions of the item, like "flammable", "metallic", etc. A lot of Overhaul's features rely on such system, so I definitely wouldn't mind implementing said system inside tML.

**Comment by hamstar0** [comment]
> @hamstar0 here's one I gave to CB: https://paste.mod.gg/jituketadi.cpp
> Tags are the same thing you meant by groups. They're descriptions of the item, like "flammable", "metallic", etc. A lot of Overhaul's features rely on such system, so I definitely wouldn't mind implementing said system inside tML.

That works if all mods had a reference to the ItemFlammableTag class, but that might not always be the case with a crowd sourced list. If a system like that were added, the group definitions would have to exist in tModLoader. Mods could only add or remove from them, or else have to expect dependencies.

Is `TagData<TagShortcut>()` meant to be a constructor call? This looks unfamiliar to me.

**Comment by Mirsario** [comment]
> That works if all mods had a reference to the ItemFlammableTag class

No, you misunderstood the concept. That class is just a local shortcut, it does not define a tag, it's like a ref to data of such tag.

> Is `TagData<TagShortcut>()` meant to be a constructor call? This looks unfamiliar to me.

The parenthesis was a typo, replaced the link.

**Comment by hamstar0** [comment]
> > That works if all mods had a reference to the ItemFlammableTag class, but that might not always be the case with a crowd sourced list. If a system like that were added, the group definitions would have to exist in tModLoader.
> 
> No, you misunderstood the concept. That class is just a local shortcut, it does not define a tag, it's like a ref to data of such tag.
> 
> > Is `TagData<TagShortcut>()` meant to be a constructor call? This looks unfamiliar to me.
> 
> The parenthesis was a typo, replaced the link.

Alright, but if I'm understanding this correctly, entity group data is mapped to that (generic) class type. The generic parameter is the class that would need to be mod-agnostic, AKA stored in tModLoader or else expected as a mod dependency.

**Comment by JavidPack** [comment]
Closing since #4381 should cover most of the desired functionality.
